decentraland/unity-renderer
# chore: Create an AppMode to manage several states of the game

### Why?
We often need to implement specific behaviours depending on different situations that happens in our game (for example: enable/unable certain buttons of the taskbar when the Builder In World editor is open/closed). Currently we are managing these cases implementing very specific logics in specifics places, making it difficult to trace and also making it difficult remember that we possibly should update this logic in future changes.

To avoid this, we would need some kind of system that let us set in any moment a "state" for the game. For example: I open the Builder In World editor, then I would like to set the game state as something that tell us "_We are executing the Builder In World editor_". In this way we would be able to know in any moment what is the game state and apply a more generic logic in the side of all those consumers that need to do something depending on specific states.

### Implementation of the `appMode` variable
We have created a new variable `appMode` in our `DataStore`. The type of this new variable is a simple enum that contains the different modes (or states) of our application. For now we have created only 2 modes (we should add new modes to the enum as we need it):
- `DEFAULT`: to define the "normal" state of the game.
- `BUILDER_IN_WORLD_EDITION`: to define when we are editing a scene with the Builder In World feature.

In the consumers side (for example: the `TaskbarButton` class) the only thing they should do is to subscribe to `DataStore.i.appMode.OnChange` and implement the needed logic (for the case of the `TaskbarButton`: enable or unable the button when we enter in specifics game modes).

### Testing
1. Go to: https://play.decentraland.zone/?index.html&renderer=urn:decentraland:off-chain:renderer-artifacts:chore/Create-an-AppMode-to-manage-several-states-of-the-game&ENV=zone&ENABLE_BUILDER_IN_WORLD&position=100,100
2. Press `K` to open the Builder In World editor.
3. Notice that the explore and tutorial buttons are been deactivated in the taskbar.
![image](https://user-images.githubusercontent.com/64659061/116700471-1886b700-a9c7-11eb-8ae4-dd5346ef7cdf.png)
4. Exit from the BiW editor.
5. Notice that the explore and tutorial buttons are been activated in the taskbar.
![image](https://user-images.githubusercontent.com/64659061/116700354-fa20bb80-a9c6-11eb-835a-982cc06116af.png)
